Update DPC++

pranav-sivaraman opened the pull request:
## Changes
- Include latest release according to the March 2023 release notes
- Support for new architectures
- Changes made are to ensure they match the [buildbot](https://github.com/intel/llvm/tree/sycl/buildbot) scripts in the original repository

## Concern
- Currently when compilng ex: `clang/clang++ -fsycl *.cpp -o *.app`. You need to link `dpcpp/bin/lib` and `gcc/<version>/lib64` to the executable. I am not sure how I would go about adding this functionality to the Spack package. I believe you have to add `-Wl rpath=`? However, what would be the best practice in adding this functionality? Would appreciate if someone can provide some input. 

- Also propose removing the [intel-llvm](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/intel-llvm/package.py) package as it redundant and also had not been updated in a while.

pranav-sivaraman committed:
refactor(dpcpp): rename rocm variants

Rename rocm and rocm-platform variants to hip and hip-platform.
New versions of DPC++ use HIP instead of ROCm.

pranav-sivaraman committed:
feat(dpcpp): add kernel fusion variant

pranav-sivaraman committed:
feat(dpcpp): add sycl enabled plugins cmake argument

pranav-sivaraman committed:
refactor(dpcpp): rename esimd-cpu variant to esimd-emulator

pranav-sivaraman committed:
refactor(dpcpp): rename shared variant to shared-libs

pranav-sivaraman committed:
refactor(dpcpp): remove openmp variant

pranav-sivaraman committed:
refactor(dpcpp): remove remangle_libclc variant

pranav-sivaraman committed:
refactor(dpcpp): update libclc_targets_to_build to use correct targets

pranav-sivaraman committed:
refactor(dpcpp): swap True to "ON" in cmake args

pranav-sivaraman committed:
feat(dpcpp): add xpti_enable_werror cmake arg

pranav-sivaraman committed:
fix(dpcpp): remove libclc project when building on macOS

pranav-sivaraman committed:
feat(dpcpp): add support for more architectures

pranav-sivaraman committed:
feat(dpcpp): add 2023-03 release notes build of DPC++

pranav-sivaraman committed:
feat(dpcpp): add security flags variant

pranav-sivaraman committed:
fix(dpcpp): set the correct build targets for install

pranav-sivaraman committed:
feat(dpcpp): add compiler properties for dpcpp

pranav-sivaraman committed:
refactor(dpcpp): simplify setting up run environment

pranav-sivaraman committed:
fix(dpcpp): add conflicts to ensure DPC++ is only built with gcc

pranav-sivaraman committed:
feat(dpcpp): add llvm external projects variant

pranav-sivaraman committed:
feat(dpcpp): remove maximum CUDA requirement

pranav-sivaraman committed:
refactor(dpcpp): replaced compiler conflict with requires

pranav-sivaraman committed:
fix(dpcpp): add conflicts for lld variant

alalazo commented:
@spackbot fix style

pranav-sivaraman committed:
fix(dpcpp): fix style

G-071 commented:
This PR is rather timely for me as I am currently working on creating a spack package for octotiger ([over here](https://github.com/G-071/octotiger-spack)) and the dpcpp package changes here make it a lot easier to add a sycl variant to it (Thanks!).

However, I encountered one small issue when using this PR: creating a spack dev-build of my own package (with a dpcpp dependency) and building it (with dpcpp) works well. However, when trying to run the resulting executable (within the dev-build shell where everthing should be loaded), I get the error
```error while loading shared libraries: libsycl.so.6: cannot open shared object file: No such file or directory``` 
until I somehow load the library manually (for example with some ugly workaround like ```
export LD_LIBRARY_PATH="$(spack find -l --paths dpcpp | tail -n1 | awk '{ print $3 }')/lib":${LD_LIBRARY_PATH}``` - a simple spack load dpcpp did not work).

I played around with this issue for a bit and found that adding the following to the dpcpp/package.py  resolves the problem by setting up the path for packages that depend on dpcpp (like mine) automatically
```py
    def setup_dependent_run_environment(self, env, dependent_spec):
        env.prepend_path("LD_LIBRARY_PATH", join_path(self.spec.prefix, "lib"))

    def setup_dependent_build_environment(self, env, dependent_spec):
        env.prepend_path("LD_LIBRARY_PATH", join_path(self.spec.prefix, "lib"))
```
Would it be possible to add this to the PR? (or something that similarly resolves the issue - I am not sure if the above would be the recommended way to tackle this at all, I just figured to include it as a possible solution since it seems to work for me)

PerryStyle commented:
@G-071 Thanks for the input! I am glad to see others having some success with the DPC++ package. Let me address your points.

> However, when trying to run the resulting executable (within the dev-build shell where everthing should be loaded), I get the error `libsycl.so.6: cannot open shared object file: No such file or directory` until I somehow load the library manually
>
Yeah I am aware of this issue when you do a `spack load dpcpp`. In order to fix that error `intel/llvm` docs actually recommend you do an `export LD_LIBRARY_PATH=$DPCPP_HOME/llvm/build/lib:$LD_LIBRARY_PATH`. So your proposed changes would be suitable into fixing this. 

However, I think the Spack solutions wants to avoid modifying any environment variables like that and instead opt for doing rpath linking. So you would do `clang -fsycl -Wl;-rpath=llvm/build/bin app`. Kind of what the `intel-oneapi-compilers` package is doing. Currently trying to implement it but that make take a while.

You can also use `DPC++` as the compiler in your package spec. Testing it on a toy project using `%clang` which points to the `DPC++` install seems to link everything just fine. Although, I am not sure how it would work with a spack environment file so the above might need to be implemented.

pranav-sivaraman committed:
refactor(dpcpp): deprecate old versions

rscohn2 commented:
> * You need to link `dpcpp/bin/lib` and `gcc/<version>/lib64` to the executable.

Are you using the compiler by doing:
```
spack load dpcpp
```
and then invoking the compiler outside of spack? If so, then you could add a `setup_run_environment` that defines LD_LIBRARY_PATH, and any other env variables that you would need to set to make the compiler work.

pranav-sivaraman committed:
fix(dpcpp): prepend DPC++ library paths to LD_LIBRARY_PATH

PerryStyle commented:
@spackbot fix style

Commit:
[@spackbot] updating style on behalf of PerryStyle

pranav-sivaraman committed:
fix(dpcpp): add hip depedency when building hip backend

pranav-sivaraman committed:
fix(dpcpp): hip backend depends on rocm

pranav-sivaraman committed:
fix(dpcpp): hip dependency

pranav-sivaraman committed:
fix(dpcpp): enable lld project when +lld

pranav-sivaraman committed:
feat: remove gcc requirement

pranav-sivaraman commented:
@spackbot fix style

pranav-sivaraman committed:
chore: fix style

pranav-sivaraman committed:
Merge branch 'develop' into packages/dpcpp-update

alalazo commented on var/spack/repos/builtin/packages/dpcpp/package.py:
```diff
@@ -16,80 +17,122 @@ class Dpcpp(CMakePackage):
+
     variant("cuda", default=False, description="switch from OpenCL to CUDA")
-    variant("rocm", default=False, description="switch from OpenCL to ROCm")
+    variant("hip", default=False, description="switch from OpenCL to HIP")
```
Why changing name to a variant that is widely used throughout Spack?

pranav-sivaraman replied:
I changed the name to HIP since DPC++ supports both HIP AMD and HIP NVIDIA backends. Also the [buildbot](https://github.com/intel/llvm/blob/sycl/buildbot/configure.py) script I used as a reference renamed it from rocm to hip. 

Although, DPC++ recommends not to build it with HIP NVIDIA and just prefer CUDA. So we could change it back to rocm if its preferred.

sethrj replied:
HIP nvidia is unfortunately a joke. It was generated by AMD to try to make their compiler seem cross platform, but in reality every scientific app has to write another compatibility layer of their own to work with HIP and CUDA. Not even Cmake supports using the nvidia backend for hip. It's best to look at HIP as a cuda like layer that AMD added to rocm. Please don't change the variant name.

sethrj replied:
https://docs.olcf.ornl.gov/systems/summit_user_guide.html#hip-code-cannot-be-built-using-cmake-using-hip-host-device-or-hip-language-support

pranav-sivaraman replied:
Yeah from looking at other scientific applications it seems like a common thing to just avoid HIP NVIDIA. I'll change back to the rocm variant.

pranav-sivaraman replied:
If I change the variant back to `rocm`, what should I do with the `hip-platform` variant? Should I assume that we always we want to build the HIP AMD backend and ignore the HIP NVIDIA backend?

alalazo commented on var/spack/repos/builtin/packages/dpcpp/package.py:
```diff
@@ -16,80 +17,122 @@ class Dpcpp(CMakePackage):
     depends_on("ninja@1.10.0:", type="build")
 
-    depends_on("cuda@10.2.0:11.4.999", when="+cuda")
+    depends_on("cuda@10.2.0:", when="+cuda")
```
Can we reuse anything from the `CudaPackage` and `RocmPackage` base classes?

pranav-sivaraman replied:
Potentially could be useful use the `CudaPackage` and `RocmPackage`. Although I don't think there is any use of the `cuda_arch` and `amdgpu_target` variants as we are simply just building the backend. I think `RocmPackage` also enforces `amdgpu_target` by default. 

Also the `intel/llvm` documentation recommends to use at least CUDA 11.6 as their were issues some math built-ins when using the optimization flags. Is there any way we can recommend this to the concertizer? Or at least show up when doing a `spack info`.

sethrj commented on var/spack/repos/builtin/packages/dpcpp/package.py:
```diff
@@ -101,45 +144,59 @@ def cmake_args(self):
             self.define("LLVM_ENABLE_PROJECTS", llvm_enable_projects),
             self.define("LIBCLC_TARGETS_TO_BUILD", libclc_targets_to_build),
-            self.define_from_variant("SYCL_BUILD_PI_CUDA", "cuda"),
-            self.define_from_variant("SYCL_BUILD_PI_ROCM", "rocm"),
```
Are all these changes compatible with older DPC++ versions? I'm guessing not: you should find what version the internals changed from using ROCM to HIP (for example) and add logic here so that older versions will still compile. Then when the "deprecated" versions are removed you can delete them.

pranav-sivaraman committed:
feat: relax build requirements

pranav-sivaraman committed:
refactor: self.spec to spec

pranav-sivaraman committed:
refactor: switch hip varaint to rocm

pranav-sivaraman committed:
fix: libclc gpu target names

pranav-sivaraman committed:
refactor: gpu plugin dependencies

pranav-sivaraman committed:
fix: style

pranav-sivaraman committed:
fix: refactor hip to rocm

pranav-sivaraman committed:
fix: amd libclc target name

pranav-sivaraman committed:
feat: enforce ninja build system

pranav-sivaraman committed:
feat: add 2023-10 release

pranav-sivaraman committed:
refactor: develop version to sycl

pranav-sivaraman committed:
refactor: remove cuda/rocm variants in favor of mixin classes

pranav-sivaraman committed:
refactor: remove llvm-external-projects

pranav-sivaraman committed:
refactor: remove cuda/hip dependencies since they are inherited

pranav-sivaraman committed:
Merge branch 'develop' into packages/dpcpp-update

pranav-sivaraman committed:
fix: add build options for old variants

pranav-sivaraman committed:
refactor: use llvm_targets_to_build function from llvm package

pranav-sivaraman committed:
feat: add llvm variants from llvm package

pranav-sivaraman committed:
fix: set version guards for variants and build requirements

pranav-sivaraman committed:
style: fix style

pranav-sivaraman committed:
refactor: set lld variant to true

pranav-sivaraman committed:
fix: add gcc install prefix option

pranav-sivaraman committed:
fix: sha256 for 2023-10

pranav-sivaraman committed:
revert: llvm targets to build

pranav-sivaraman committed:
fix: don't set default value lld to true

pranav-sivaraman committed:
refactor: remove LLVM targets

Some reason intel/llvm doesn't support building with the targets all option
even though it syncs with upstream

pranav-sivaraman committed:
style: isort style fixes

pranav-sivaraman committed:
fix: remove variant that does not exist

pranav-sivaraman committed:
feat: add v2023-11

pranav-sivaraman committed:
fix: get /opt/rocm/ directory instead of /opt/rocm/hip/

pranav-sivaraman committed:
Revert "fix: get /opt/rocm/ directory instead of /opt/rocm/hip/"

This reverts commit a991b6aed9defebf9067b1f0e7ecb193a4eca987.

pranav-sivaraman commented:
Closing this PR since the DPC++ compiler is constantly changing and prone to bugs.  It makes more sense to just build manually on either develop or a nightly release.
